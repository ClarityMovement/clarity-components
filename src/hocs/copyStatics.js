import { compose, setStatic, getDisplayName } from 'recompose';

export default SrcComponent => compose(
  setStatic('displayName', getDisplayName(SrcComponent)),
  setStatic('propTypes', SrcComponent.propTypes),
  setStatic('defaultProps', SrcComponent.defaultProps),
);
