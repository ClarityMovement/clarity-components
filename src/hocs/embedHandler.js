import withHandlers from 'recompose/withHandlers';

export const createEmbeddedFunction = (innerFunc, outerFunc) => (...innerArgs) => {
  if (outerFunc.length <= innerFunc.length) {
    innerFunc(...innerArgs);
    return outerFunc(...innerArgs);
  }
  const next = () => innerFunc(...innerArgs);
  const outerArgs = [...innerArgs];
  outerArgs.length = innerFunc.length;
  return outerFunc(...outerArgs, next);
};

export const createEmbeddedHandler = (innerName, outerName) => (props) => {
  let innerHandler;
  if (typeof innerName === 'function') {
    innerHandler = innerName(props);
  } else if (typeof innerName === 'string') {
    innerHandler = props[innerName];
  }
  if (typeof innerHandler !== 'function') {
    throw new Error('innerName must be a handler or the name of it');
  }

  const outerHandler = props[outerName];

  if (!outerHandler) return innerHandler;

  return createEmbeddedFunction(innerHandler, outerHandler, props);
};

export default (innerName, outerName) => withHandlers({
  [outerName]: createEmbeddedHandler(innerName, outerName),
});

