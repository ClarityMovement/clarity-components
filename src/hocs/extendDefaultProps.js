import setStatic from 'recompose/setStatic';

export default defaultProps => (Component) => {
  if (!defaultProps) return Component;

  return setStatic('defaultProps', {
    ...Component.defaultProps,
    ...defaultProps,
  })(Component);
};
