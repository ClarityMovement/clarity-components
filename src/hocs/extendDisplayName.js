import setDisplayName from 'recompose/setDisplayName';

// utils
import getDisplayName from 'recompose/getDisplayName';
import wrapDisplayName from 'recompose/wrapDisplayName';

export default nameOrOptions => (Component) => {
  const options = typeof nameOrOptions !== 'object' ? { name: nameOrOptions } : nameOrOptions;
  const { name, operation = 'wrap' } = options;

  let newDisplayName;
  switch (operation) {
    case 'wrap': {
      newDisplayName = wrapDisplayName(Component, name);
      break;
    }
    case 'replace': {
      newDisplayName = name;
      break;
    }
    default : {
      newDisplayName = getDisplayName(Component);
    }
  }
  return setDisplayName(newDisplayName)(Component);
};
