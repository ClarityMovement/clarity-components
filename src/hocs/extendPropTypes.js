import setPropTypes from 'recompose/setPropTypes';

export default propTypes => (Component) => {
  if (!propTypes) return Component;

  return setPropTypes({
    ...Component.propTypes,
    ...propTypes,
  })(Component);
};
