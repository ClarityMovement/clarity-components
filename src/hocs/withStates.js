import _ from 'lodash';
import { compose, withReducer, withHandlers, flattenProp } from 'recompose';

import { omitProps } from 'react-render-counter/hocs';

const MERGE_STATE = 'MERGE_STATE';
const RESET_STATE = 'RESET_STATE';

export const reducer = (state, { type, payload } = {}) => {
  switch (type) {
    case MERGE_STATE: {
      return { ...state, ...payload };
    }
    case RESET_STATE: {
      return { ...payload };
    }
    default:
  }
  return state;
};

export const createHandlers = (stateKeys, dispatchName) => {
  const handlers = {};
  stateKeys.forEach((stateKey) => {
    const handler = ({ [dispatchName]: setState }) => value => setState({
      type: MERGE_STATE,
      payload: { [stateKey]: value },
    });
    handlers[`set${_.capitalize(stateKey)}`] = handler;
  });
  return handlers;
};

export const createResetState = (initialState, dispatchName) =>
  ({ [dispatchName]: dispatch, ...props }) => (newState = initialState) => dispatch({
    type: RESET_STATE,
    payload: (typeof newState === 'function' ? newState(props) : newState),
  });

export const createSetState = dispatchName =>
  ({ [dispatchName]: dispatch }) => newState => dispatch({
    type: MERGE_STATE,
    payload: newState,
  });

export const createWithStatesHocs = (initialState, {
  stateKeys,
  names: {
    state: stateName,
    dispatch: dispatchName,
    merge: mergeName,
    reset: resetName,
  },
  omit,
}) => {
  const handlers = createHandlers(stateKeys, dispatchName);
  handlers[resetName] = createResetState(initialState, dispatchName);
  handlers[mergeName] = createSetState(dispatchName);

  const hocs = [
    withReducer(stateName, dispatchName, reducer, initialState),
    withHandlers(handlers),
    flattenProp(stateName),
  ];

  const propsToOmit = _.keys(_.pickBy(omit, _.truthy));
  if (propsToOmit.length) {
    hocs.push(omitProps(propsToOmit));
  }

  return hocs;
};

export const createGetInitialState = (initialKeysOrState) => {
  const type = typeof initialKeysOrState;
  if (type !== 'object' && type !== 'function') {
    throw new Error('initialKeysOrState must be an object or a function');
  }

  if (Array.isArray(initialKeysOrState)) {
    const initialState = {};
    initialKeysOrState.forEach((key) => {
      initialState[key] = undefined;
    });
    return initialState;
  }
  return initialKeysOrState;
};

const defaultOptions = {
  names: {
    state: 'state',
    dispatch: 'dispatchState',
    reset: 'resetState',
    merge: 'setState',
  },
  omit: {
    dispatch: true,
    state: true,
    reset: false,
    merge: false,
  },
};

export const createOptions = (initialState, options = {}) => {
  if (typeof initialState === 'function' && !options.stateKeys) {
    throw new Error('options.stateKeys must be specified when initialState is a function');
  }

  const mergedOptions = _.merge({}, defaultOptions, options);

  const stateKeys = options.stateKeys || Object.keys(initialState);

  const omit = _.mapKeys(mergedOptions.omit, (value, key) => mergedOptions.names[key]);

  return {
    ...mergedOptions,
    stateKeys,
    omit,
  };
};


/*
 * Compatible for different params
 * withStates(
 *   initialKeys: Array<string> |
 *   initialState: {
 *     [stateKey: string]: initialValue: Any
 *   } |
 *   getInitialState: (props) => {
 *     [stateKey: string]: initialValue: Any
 *   },
 *   options: {
 *     stateName: string,
 *     dispatchName: string,
 *     stateKeys: Array<string>,
 *   },
 * )
 *
 * TODO: withStates(
 *   initialState: {
 *     [stateKey: string]: getInitialValue: Function
 *   },
 *   options: Object
 * )
 */
export default (
  initialKeysOrState,
  options,
) => {
  const initialState = createGetInitialState(initialKeysOrState);
  const finalOptions = createOptions(initialState, options);

  const hocs = createWithStatesHocs(initialState, finalOptions);
  return compose(...hocs);
};
