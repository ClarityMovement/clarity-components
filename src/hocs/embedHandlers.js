import _ from 'lodash';
import { createEmbeddedHandler, createEmbeddedName } from './embedHandler';
import withHandlers from './withHandlers';

const createEmbeddedHandlers = (handlers) => {
  const embeddedHandlers = {};

  _.forEach(handlers, (innerName, outerName) => {
    const name = createEmbeddedName(innerName, outerName);
    embeddedHandlers[name] = createEmbeddedHandler(innerName, outerName);
  });

  return embeddedHandlers;
};

export default (handlers) => {
  const handlersArray = Array.isArray(handlers) ? handlers : [handlers];
  return withHandlers(handlersArray.map(createEmbeddedHandlers));
};
