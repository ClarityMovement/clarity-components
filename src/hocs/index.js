import copyStatics from './copyStatics';
import embedHandler, { createEmbeddedFunction, createEmbeddedHandler } from './embedHandler';
import embedHandlers from './embedHandlers';
import extendDefaultProps from './extendDefaultProps';
import extendDisplayName from './extendDisplayName';
import extendPropTypes from './extendPropTypes';
import extendStatics from './extendStatics';
import omitProps from './omitProps';
import omitPropTypes from './omitPropTypes';
import withHandlers from './withHandlers';
import withPropsPeeker from './withPropsPeeker';
import withThis from './withThis';

export {
  copyStatics,
  embedHandler,
  createEmbeddedHandler,
  createEmbeddedFunction,
  embedHandlers,
  extendDefaultProps,
  extendDisplayName,
  extendPropTypes,
  extendStatics,
  omitProps,
  omitPropTypes,
  withHandlers,
  withPropsPeeker,
  withThis,
};
