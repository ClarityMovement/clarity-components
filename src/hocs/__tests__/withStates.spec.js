/* eslint-env jest */
import React from 'react';
import { mount } from 'enzyme';
import { withPropsPeeker } from 'react-render-counter/hocs';
import snapshotHocProps from '../../utils/testHelpers/snapshotHocProps';
import withStates, { MERGE_STATE, RESET_STATE } from '../withStates';

describe('withStates(initialKeysOrState, options): PropsMapperHoc', () => {
  const initialState = {
    foo: 1,
    bar: 2,
  };
  const stateKeys = Object.keys(initialState);

  describe('when initialKeysOrState is an object as initialState', () => {
    test('with default options', () => {
      snapshotHocProps(withStates(initialState));
    });

    test('with options.stateKeys', () => {
      snapshotHocProps(withStates(initialState, {
        stateKeys: ['foo', 'bar', 'baz'],
      }));
      snapshotHocProps(withStates(initialState, {
        stateKeys: ['foo'],
      }));
    });
  });

  describe('when initialKeysOrState is an array of keys of initialState', () => {
    test('with default options', () => {
      snapshotHocProps(withStates(stateKeys));
    });
  });

  describe('when initialKeysOrState is a function to create initialState', () => {
    const createInitialState = () => initialState;
    test('with default options', () => {
      const stateKeysRequired = 'options.stateKeys must be specified when initialState is a function';
      expect(() => withStates(createInitialState)).toThrow(stateKeysRequired);
    });

    test('with options.stateKeys', () => {
      snapshotHocProps(withStates(createInitialState, { stateKeys }));
    });
  });

  describe('when initialKeysOrState is invalid', () => {
    const typeCheck = 'initialKeysOrState must be an object or a function';
    test(typeCheck, () => {
      expect(() => withStates(2)).toThrow(typeCheck);
    });
  });

  describe('output props have some meta handlers', () => {
    test('with options.omit to hide them', () => {
      const hideAll = { dispatch: true, merge: true, state: true, reset: true };
      snapshotHocProps(withStates(initialState, { omit: hideAll }));
    });

    const showAll = { dispatch: false, merge: false, state: false, reset: false };
    test('with options.omit to show them', () => {
      snapshotHocProps(withStates(initialState, { omit: showAll }));
    });

    test('with options.names to rename them', () => {
      const names = {
        dispatch: 'dispatch',
        merge: 'merge',
        state: 'finalState',
        reset: 'reset',
      };
      snapshotHocProps(withStates(initialState, { names, omit: showAll }));
    });

    describe('meta handlers can change states', () => {
      const newState = { foo: 3 };
      let props;

      beforeEach(() => {
        props = {};
        const BaseComponent = withPropsPeeker(props)(() => <div />);
        const NewComponent = withStates(initialState, { omit: showAll })(BaseComponent);
        mount(<NewComponent />);
      });

      test('dispatchState(action: { type, payload })', () => {
        expect(props.dispatchState).not.toBeUndefined();

        let [curState, lastState] = [props.state, undefined];
        expect(curState).toBe(initialState);
        expect(curState).not.toBe(lastState);

        props.dispatchState({
          type: MERGE_STATE,
          payload: newState,
        });
        [curState, lastState] = [props.state, curState];
        expect(curState).not.toBe(initialState);
        expect(curState).not.toBe(lastState);
        expect(curState).toEqual({
          foo: 3,
          bar: 2,
        });

        props.dispatchState(undefined);
        [curState, lastState] = [props.state, curState];
        expect(curState).toBe(lastState);

        props.dispatchState({
          type: RESET_STATE,
          payload: newState,
        });
        [curState, lastState] = [props.state, curState];
        expect(curState).not.toBe(lastState);
        expect(curState).toEqual(newState);
      });

      test('resetState(newState)', () => {
        expect(props.resetState).not.toBeUndefined();

        expect(props.state).toEqual(initialState);
        props.resetState(newState);
        expect(props.state).toEqual(newState);
        props.resetState();
        expect(props.state).toEqual(initialState);
        props.resetState(() => newState);
        expect(props.state).toEqual(newState);
      });

      test('setState(newState)', () => {
        expect(props.setState).not.toBeUndefined();

        expect(props.state).toEqual(initialState);
        props.setState(newState);
        expect(props.state).toEqual({ foo: 3, bar: 2 });
      });

      test('[setStateKey](newStateValue)', () => {
        expect(props.setFoo).not.toBeUndefined();

        expect(props.foo).toBe(1);
        props.setFoo(newState.foo);
        expect(props.foo).toBe(newState.foo);
      });
    });
  });
});
