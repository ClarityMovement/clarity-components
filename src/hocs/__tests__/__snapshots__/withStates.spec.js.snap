// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withStates(initialKeysOrState, options): PropsMapperHoc output props have some meta handlers with options.names to rename them 1`] = `
<div
  name="props-displayer"
>
  <input-props />
  <output-props
    bar={2}
    dispatch={[Function]}
    finalState={
      Object {
        "bar": 2,
        "foo": 1,
      }
    }
    foo={1}
    merge={[Function]}
    reset={[Function]}
    setBar={[Function]}
    setFoo={[Function]}
  />
</div>
`;

exports[`withStates(initialKeysOrState, options): PropsMapperHoc output props have some meta handlers with options.omit to hide them 1`] = `
<div
  name="props-displayer"
>
  <input-props />
  <output-props
    bar={2}
    foo={1}
    setBar={[Function]}
    setFoo={[Function]}
  />
</div>
`;

exports[`withStates(initialKeysOrState, options): PropsMapperHoc output props have some meta handlers with options.omit to show them 1`] = `
<div
  name="props-displayer"
>
  <input-props />
  <output-props
    bar={2}
    dispatchState={[Function]}
    foo={1}
    resetState={[Function]}
    setBar={[Function]}
    setFoo={[Function]}
    setState={[Function]}
    state={
      Object {
        "bar": 2,
        "foo": 1,
      }
    }
  />
</div>
`;

exports[`withStates(initialKeysOrState, options): PropsMapperHoc when initialKeysOrState is a function to create initialState with options.stateKeys 1`] = `
<div
  name="props-displayer"
>
  <input-props />
  <output-props
    bar={2}
    foo={1}
    resetState={[Function]}
    setBar={[Function]}
    setFoo={[Function]}
    setState={[Function]}
  />
</div>
`;

exports[`withStates(initialKeysOrState, options): PropsMapperHoc when initialKeysOrState is an array of keys of initialState with default options 1`] = `
<div
  name="props-displayer"
>
  <input-props />
  <output-props
    resetState={[Function]}
    setBar={[Function]}
    setFoo={[Function]}
    setState={[Function]}
  />
</div>
`;

exports[`withStates(initialKeysOrState, options): PropsMapperHoc when initialKeysOrState is an object as initialState with default options 1`] = `
<div
  name="props-displayer"
>
  <input-props />
  <output-props
    bar={2}
    foo={1}
    resetState={[Function]}
    setBar={[Function]}
    setFoo={[Function]}
    setState={[Function]}
  />
</div>
`;

exports[`withStates(initialKeysOrState, options): PropsMapperHoc when initialKeysOrState is an object as initialState with options.stateKeys 1`] = `
<div
  name="props-displayer"
>
  <input-props />
  <output-props
    bar={2}
    foo={1}
    resetState={[Function]}
    setBar={[Function]}
    setBaz={[Function]}
    setFoo={[Function]}
    setState={[Function]}
  />
</div>
`;

exports[`withStates(initialKeysOrState, options): PropsMapperHoc when initialKeysOrState is an object as initialState with options.stateKeys 2`] = `
<div
  name="props-displayer"
>
  <input-props />
  <output-props
    bar={2}
    foo={1}
    resetState={[Function]}
    setFoo={[Function]}
    setState={[Function]}
  />
</div>
`;
